### PR TITLE
fix: merge collection element schemas with differing key sets

### PR DESCRIPTION
**Pull request by ogawa-takeshi**
## Summary

Fixes a `DataException` raised by `CompactValueConverter.value()` when a Cypher `collect()` produces Maps with different key sets across elements (issue [#527](https://github.com/neo4j/neo4j-kafka-connector/issues/527)).

When the Collection branch sees multiple distinct element Schemas, `CompactValueConverter.schema()` previously fell back to the indexed-struct `{e0, e1, ...}` representation, but `value()` then could not coerce the elements and threw. This change merges those element schemas into a single STRUCT covering the union of keys; keys absent (or null/empty) in some elements become optional fields. Conflicting non-null types still fall back to the indexed-struct representation.

This is the **Case 1** fix referred to in the discussion on [#528](https://github.com/neo4j/neo4j-kafka-connector/pull/528). Per @fbiville's request, Case 1 (different key sets) and Case 2 (same key, null vs concrete value) have been split. Case 2 lives in [#528](https://github.com/neo4j/neo4j-kafka-connector/pull/528); this PR is Case 1 only.

## Addressing the "skipping keys" feedback

@fbiville pointed out on [#528](https://github.com/neo4j/neo4j-kafka-connector/pull/528) that an earlier Case 1 implementation silently dropped keys: tests that schema-shape-asserted but didn't call `converter.value(schema, coll)` masked the bug. The earlier `mergeMapAndStructSchemas` filtered `Set<Schema>` down to STRUCT schemas only and discarded MAP schemas, losing any key that lived exclusively under a MAP-typed element.

This rewrite avoids that class of bug by working from the **raw Map values**, not from previously-computed Schemas:

```kotlin
private fun mergeMapElementSchemas(elements: List<Map<*, *>>, optional: Boolean): Schema? {
  // walk every key from every element, re-inferring value schemas with
  // forceMapsAsStruct=true so MAP and STRUCT element shapes are comparable
}
```

Every new test in `CompactValueConverterTest` calls `converter.value(schema, coll)` and asserts on the resulting `List<Struct>`, so a key-dropping regression cannot pass. Specifically, `collection mixing homogeneous map and heterogeneous map should not drop keys` exercises the exact scenario that broke earlier (one element produces a `MAP<String,String>` schema, the other a STRUCT) and asserts that all three keys (`x`, `y`, `z`) round-trip.

## Scope notes

- The merge only runs when **all** non-empty Collection elements are Maps. Anything else (mixed types, conflicting non-null field types) still falls back to the existing indexed-struct schema. No behavior change for non-Map collections.
- Nested Maps with differing key sets are merged recursively (also driven by raw values, so the same safety property holds).
- @fbiville also noted the team is cautious about scanning all data for schema inference and may defer broader changes to the next major version. This PR is intentionally minimal — it scans only the immediate elements of a Collection that already hit the `else`-branch fallback (i.e. `schema()` was already iterating them anyway). If that's still beyond what should land before the next major, please feel free to defer this PR.

## Test plan

- [x] `./mvnw -pl common -am test -Dtest='DynamicTypesCompactTest'` — 28 tests pass (21 pre-existing + 7 new)
- [x] Updated `Neo4jSourceQueryIT.serializes list of heterogeneous objects as merged list by default` to reflect the new merged-list output
- [ ] Reviewer to run full source-connector ITs locally

### CODE REVIEW

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 1]
```diff
@@ -703,4 +703,162 @@ class DynamicTypesCompactTest {
...
+
+    schema.type() shouldBe Schema.Type.ARRAY
+    val element = schema.valueSchema()
+    element.type() shouldBe Schema.Type.STRUCT
+    element.fields().map { it.name() }.toSet() shouldBe setOf("name", "age", "address")
+    element.field("name").schema().isOptional shouldBe false
+    element.field("age").schema().isOptional shouldBe false
+    element.field("address").schema().isOptional shouldBe true
+
+    val addressSchema = element.field("address").schema()
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:729]: this should be replaced with a manually reconstructed schema imo

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 2]
```diff
@@ -703,4 +703,162 @@ class DynamicTypesCompactTest {
...
+                "name" to "jane",
+                "address" to mapOf("city" to "paris", "zip" to 75001, "country" to "fr"),
+            ),
+        )
+    val schema = converter.schema(coll, false)
+    val converted = converter.value(schema, coll)
+
+    schema.type() shouldBe Schema.Type.ARRAY
+    val element = schema.valueSchema()
+    val addressSchema = element.field("address").schema()
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:824]: same comment as above

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 3]
```diff
@@ -703,4 +703,162 @@ class DynamicTypesCompactTest {
                 mapOf("2000" to "birth", "2005" to "school", "2017" to "college"),
         )
   }
+
+  @Test
+  fun `collection of maps with differing key sets should merge into array of struct`() {
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:708]: nit: I'd replace "differing" with something more specific
this test focuses on the 2nd key set being a super set of the first one (I'm just not sure of a concise way to describe this 😅)

Comment (fbiville) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:708]: I'd also like to see a test where none of the keys across list elements are common

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 4]
```diff
@@ -703,4 +703,162 @@ class DynamicTypesCompactTest {
...
+            Struct(element).put("name", "john").put("age", 21L).put("address", null),
+            Struct(element)
+                .put("name", "jane")
+                .put("age", 25L)
+                .put("address", Struct(addressSchema).put("city", "london").put("zip", 10001L)),
+        )
+  }
+
+  @Test
+  fun `collection mixing homogeneous map and heterogeneous map should not drop keys`() {
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:741]: can you clarify the test name here?

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 5]
```diff
@@ -703,4 +703,162 @@ class DynamicTypesCompactTest {
...
+  fun `collection of maps with conflicting non-null types should fall back to indexed struct`() {
+    val coll = listOf(mapOf("name" to "john"), mapOf("name" to 42))
+    val schema = converter.schema(coll, false)
+
+    schema.type() shouldBe Schema.Type.STRUCT
+    schema.fields().map { it.name() } shouldBe listOf("e0", "e1")
+  }
+
+  @Test
+  fun `collection mixing maps and non-maps should fall back to indexed struct`() {
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:801]: another similar test to add are maps with non-string keys

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt [hunk 6]
```diff
@@ -304,4 +318,66 @@ class CompactValueConverter : ValueConverter {
...
+    val builder = SchemaBuilder.struct()
+    for (key in allKeys) {
+      val fieldValues =
+          elements.filter { it.containsKey(key) && it[key].notNullOrEmpty() }.map { it[key] }
+
+      val resolvedSchema =
+          if (fieldValues.isEmpty()) {
+            SimpleTypes.NULL.schema(true)
+          } else {
+            val fieldSchemas = fieldValues.map { schema(it, optional, true) }.toSet()
```
Comment (fbiville) [inline common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt:350]: question: is it necessary to force maps as struct here?

**Review by fbiville — COMMENTED** [review]
thanks for your patience, the PR mostly looks good

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt [hunk 7]
```diff
@@ -304,4 +318,66 @@ class CompactValueConverter : ValueConverter {
...
+              fieldSchemas.size == 1 -> fieldSchemas.first()
+              fieldValues.all { it is Map<*, *> } -> {
+                @Suppress("UNCHECKED_CAST")
+                mergeMapElementSchemas(fieldValues as List<Map<*, *>>, optional) ?: return null
+              }
+              else -> return null
+            }
+          }
+
+      val presentInAll = elements.all { it.containsKey(key) && it[key].notNullOrEmpty() }
```
Comment (fbiville) [inline common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt:361]: suggestion: to maximize schema compatibility, we should make the schema optional in all cases (maybe a future message would miss some fields)

**Comment by fbiville** [comment]
We'll prioritize merging https://github.com/neo4j/neo4j-kafka-connector/pull/528.
Then, we _may_ follow up with this PR. If that happens, we will probably guard the behavior the PR introduces with a new opt-in configuration setting.

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt (hunk 7 above)
Comment (ogawa-takeshi) [reply common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt]: Done in bd28a59d. Every field of the merged struct is now optional regardless of whether it is present in all elements, matching the change made on #528. The merged struct itself still follows the `optional` argument, mirroring how single-typed collection elements are handled. The tests were updated to assert the fully reconstructed schema.

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt (hunk 6 above)
Comment (ogawa-takeshi) [reply common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt]: Good question — it turned out not to be necessary, so I removed it in bd28a59d. I had added it so that a key inferred as `MAP` in one element and as `STRUCT` in another would be comparable, but the recursive merge already handles that case since it works from the raw values. The merge now propagates the caller's `forceMapsAsStruct` flag unchanged, so nested single-typed maps keep their `MAP` schema as elsewhere. Added `nested maps with single typed values should keep their map schema when merging` to cover this.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 1 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt]: Done in bd28a59d. All the new tests now build the expected schema by hand with `SchemaBuilder` and assert `schema shouldBe SchemaBuilder.array(expectedElementSchema).build()` instead of reading sub-schemas back from the result.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 2 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt]: Same as above — done in bd28a59d, the expected `address` schema is now reconstructed manually.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 3 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt]: Renamed in bd28a59d to `collection of maps where one key set is a superset of the other should merge into array of struct`.

Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt]: Added `collection of maps with no common keys should merge into array of struct` in bd28a59d: `[{name: "john"}, {age: 21}]` merges into a struct with both fields optional, and `value()` round-trips both elements.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 4 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt]: Renamed in bd28a59d to `collection of a map inferred as MAP schema and a map inferred as STRUCT schema should keep all keys`, and reworded the comment to explain that `a` (all string values) is inferred as `MAP<STRING, STRING>` on its own while `b` (mixed value types) is inferred as a `STRUCT`, and that the merge must not lose `y`, which only exists in the MAP-typed element.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 5 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:831]: Added `collection of maps with differing key sets should still enforce string map keys` in bd28a59d: non-String keys are rejected by `schema()` before the merge is reached, so the behaviour is unchanged and it throws `IllegalArgumentException("unsupported map key type java.lang.Integer")`. I also removed the now-unreachable `key !is String` check from the merge and documented that precondition in the KDoc.

**Comment by ogawa-takeshi** [comment]
Thanks. #528 has been updated for your latest suggestions, and this PR is now rebased on `main` (bd28a59d) with all the review comments above addressed: fields of the merged struct are always optional (same as #528), the `forceMapsAsStruct` forcing was dropped since the recursive merge already covers MAP/STRUCT mixes, and the tests now assert hand-built schemas with the additional cases you asked for (no common keys, non-String map keys).

If you would like this behaviour behind an opt-in setting, happy to add it here — just let me know the preferred config key, or feel free to handle it on your side when you pick this up.

**Comment by venikkin** [comment]
Hi @ogawa-takeshi. 
Thank you for the updates. I've merged a [rebased version of your PR ](https://github.com/neo4j/neo4j-kafka-connector/pull/745). Closing this PR.
